### PR TITLE
LSP: Hover, go to declaration, and code lens implementation for Groovy

### PR DESCRIPTION
**Pull request by dbalek**
* LSP: Groovy Spock test methods finder implemented
* LSP: Hover and go to declaration implementated for Groovy

### CODE REVIEW

File: ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/spi/ComputeTestMethods.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public void cancel();
+
+    /**
+     * Factory for creating {@link ComputeTestMethods} instances.
+     * It should be registered in the {@link MimeLookup}.
+     *
+     * @since 1.25
+     */
+    public interface Factory {
```
Comment (JaroslavTulach) [inline ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/spi/ComputeTestMethods.java:56]: Interface, a factory & co. seems a bit too verbose. I would prefer interface with a single method `computeTestMethods` and some parameter:
```java
interface ComputeTestMethods {
  public void computeTestMethods(Info info);
  public final class Info {
     private Info()
     public Parser.Result getResult();
     public boolean addMethod(TestMethod m); // if returns false, consider the request cancelled
  }
}
```
with this approach the implementors need to implement just a single interface with a single method. Only one instance needs to be created. Evolution plan: add new methods into the `Info` class.

File: groovy/libs.groovy/nbproject/project.xml [hunk 2]
```diff
@@ -44,6 +44,7 @@
                 <package>org.codehaus.groovy.control</package>
                 <package>org.codehaus.groovy.control.customizers</package>
                 <package>org.codehaus.groovy.control.customizers.builder</package>
+                <package>org.codehaus.groovy.control.io</package>
```
Comment (JaroslavTulach) [inline groovy/libs.groovy/nbproject/project.xml:47]: This change should be associated with some versioning. Increment spec version of the `libs.groovy` module; make sure those who depend on the `org.codehaus.groovy.control.io` package require the new version.

File: ide/gsf.testrunner.ui/apichanges.xml [hunk 3]
```diff
@@ -64,6 +64,19 @@
...
     </change>
+    <change id="ComputeTestMethodsAndTestMethodGetTestClassPosition">
+        <api name="CommonTestrunnerUIAPI"/>
+        <summary>Added ComputeTestMethods SPI and getTestClassPosition() to TestMethod</summary>
+        <version major="1" minor="25"/>
+        <date day="26" month="6" year="2021"/>
+        <author login="dbalek"/>
+        <compatibility addition="yes"/>
+        <description>
+            Added SPI to provide a list of {@link TestMethod}s found in a parsed {@link Source}s.
```
Comment (JaroslavTulach) [inline ide/gsf.testrunner.ui/apichanges.xml:75]: `{@link}` is unlikely to work. Use `<a href="@TOP@/pkg/TestMethod.html">`

**Review by JaroslavTulach — APPROVED** [review]
Consider removing the `Factory` interface. Otherwise fine.

File: groovy/libs.groovy/nbproject/project.xml (hunk 2 above)
Comment (dbalek) [reply groovy/libs.groovy/nbproject/project.xml]: Removed. Already part of https://github.com/apache/netbeans/pull/3027

File: ide/gsf.testrunner.ui/apichanges.xml (hunk 3 above)
Comment (dbalek) [reply ide/gsf.testrunner.ui/apichanges.xml]: Fixed.

File: ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/spi/ComputeTestMethods.java (hunk 1 above)
Comment (dbalek) [reply ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/spi/ComputeTestMethods.java]: `Factory` interface removed.

**Pull request merged**